server/proxy/pf_server.c

```diff
@@ -0,0 +1,423 @@
+ * Copyright 2019 Kobi Mizrachi <kmizrachi18@gmail.com>
+ * Copyright 2019 Idan Freiberg <speidy@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <signal.h>
+
+#include <freerdp/freerdp.h>
+#include <freerdp/listener.h>
+
+#include <winpr/crt.h>
+#include <winpr/ssl.h>
+#include <winpr/synch.h>
+#include <winpr/string.h>
+#include <winpr/path.h>
+#include <winpr/winsock.h>
+#include <winpr/thread.h>
+
+#include <freerdp/channels/wtsvc.h>
+#include <freerdp/channels/channels.h>
+
+#include <freerdp/constants.h>
+#include <freerdp/server/rdpsnd.h>
+#include <freerdp/server/rdpgfx.h>
+
+#include "pf_server.h"
+#include "pf_common.h"
+#include "pf_log.h"
+#include "pf_config.h"
+#include "pf_client.h"
+#include "pf_context.h"
+#include "pf_input.h"
+#include "pf_update.h"
+#include "pf_rdpgfx.h"
+
+#define TAG PROXY_TAG("server")
+
+void pf_server_handle_client_disconnection(freerdp_peer* client)
+{
+	pServerContext* ps;
+	proxyData* pdata;
+	rdpContext* pc;
+	ps = (pServerContext*)client->context;
+	pc = (rdpContext*) ps->pdata->pc;
+	pdata = ps->pdata;
+	WLog_INFO(TAG, "Client %s disconnected; closing connection with server %s",
+	          client->hostname, pc->settings->ServerHostname);
+	/* Mark connection closed for sContext */
+	SetEvent(pdata->connectionClosed);
+	freerdp_abort_connect(pc->instance);
+	/* Close connection to remote host */
+	WLog_DBG(TAG, "Waiting for proxy's client thread to finish");
+	WaitForSingleObject(ps->thread, INFINITE);
+	CloseHandle(ps->thread);
+}
+
+static BOOL pf_server_parse_target_from_routing_token(freerdp_peer* client,
+        char** target, DWORD* port)
+{
+#define TARGET_MAX	(100)
+#define ROUTING_TOKEN_PREFIX "Cookie: msts="
+	const char* routing_token;
+	char* colon;
+	int len;
+	int prefix_len;
+	DWORD routing_token_length;
+	routing_token = freerdp_nego_get_routing_token(client->context, &routing_token_length);
+	prefix_len = strlen(ROUTING_TOKEN_PREFIX);
+
+	if (routing_token &&
+	    routing_token_length > prefix_len && routing_token_length < TARGET_MAX)
+	{
+		len = routing_token_length - prefix_len;
+		*target = malloc(len + 1);
+		CopyMemory(*target, routing_token + prefix_len, len);
+		*(*target + len) = '\0';
+		colon = strchr(*target, ':');
+		WLog_INFO(TAG, "Target [parsed from routing token]: %s", *target);
+
+		if (colon)
+		{
+			/* port is specified */
+			*port = strtoul(colon + 1, NULL, 10);
```

Comment:
No error checks here, types also do not match (32bit to 64bit on some platforms)